AFQ/api.py

```diff
@@ -71,135 +72,223 @@ def do_preprocessing():
 DIPY_GH = "https://github.com/dipy/dipy/blob/master/dipy/"
 
 
-def make_bundle_dict(bundle_names=BUNDLES,
-                     seg_algo="afq",
-                     resample_to=False):
-    """
-    Create a bundle dictionary, needed for the segmentation
+class BundleDict(MutableMapping):
+    def __init__(self,
+                 bundle_info=BUNDLES,
+                 seg_algo="afq",
+                 resample_to=False):
+        """
+        Create a bundle dictionary, needed for the segmentation
+
+        Parameters
+        ----------
+        bundle_names : list, optional
+            A list of the bundles to be used in this case. Default: all of them
+
+        seg_algo: One of {"afq", "reco", "reco16", "reco80"}
+            The bundle segmentation algorithm to use.
+                "afq" : Use waypoint ROIs + probability maps, as described
+                in [Yeatman2012]_
+                "reco" / "reco16" : Use Recobundles [Garyfallidis2017]_
+                with a 16-bundle set.
+                "reco80": Use Recobundles with an 80-bundle set.
+
+        resample_to : Nifti1Image or bool, optional
+            If set, templates will be resampled to the affine and shape of this
+            image. If False, no resampling will be done.
+            Default: False
+        """
+        if not (isinstance(bundle_info, dict)
+                or isinstance(bundle_info, list)):
+            raise TypeError((
+                f"bundle_info must be a dict or a list,"
+                f" currently a {type(bundle_info)}"))
+
+        if isinstance(bundle_info, dict):
+            self.bundle_names = list(bundle_info.keys())
+            self._dict = bundle_info.copy()
```

Comment:
Maybe this is causing the OOM? It looks like you are making a copy of the whole thing?

Reply:
No, here we are copying the user input (bundle_info) if it is a dict. It was the same in the old implementation. That way pyAFQ does not modify the user's passed in dictionary.

Reply:
Oh yeah - I see. Hmmm.

Reply:
Curiouser and curiouser: I thought maybe the BundleDict has a larger memory footprint than the dict implementation, but it seems to have a _smaller_ memory footprint. Baffling.

Reply:
Maybe the OOM is related to this issue somehow? https://github.com/yeatmanlab/pyAFQ/issues/674

I see some loky messages in the documentation build logs that are throwing the OOM. I wonder if MutableMapping is somehow more resistant to garbage collection than a vanilla dict.